pkg/consuldp/lifecycle.go

```diff
@@ -0,0 +1,195 @@
+
+		proxy: proxy,
+
+		errorExitCh: make(chan struct{}, 1),
+		mu:          sync.Mutex{},
+	}
+}
+
+func (m *lifecycleConfig) startLifecycleManager(ctx context.Context) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	if m.running {
+		return nil
+	}
+
+	m.logger = hclog.FromContext(ctx).Named("lifecycle")
+	m.running = true
+	go func() {
+		<-ctx.Done()
+		m.stopLifecycleServer()
+	}()
+
+	// Start the server which will expose HTTP endpoints for proxy lifecycle
+	// management control
+	mux := http.NewServeMux()
+
+	// Determine what HTTP endpoint paths to configure for the proxy lifecycle
+	// management server bind port is. These can be set as flags.
+	cdpLifecycleShutdownPath := defaultLifecycleShutdownPath
+	if m.gracefulShutdownPath != "" {
+		cdpLifecycleShutdownPath = m.gracefulShutdownPath
+	}
+
+	// Set config to allow introspection of default path for testing
+	m.gracefulShutdownPath = cdpLifecycleShutdownPath
+
+	fmt.Printf("setting graceful shutdown path: %s\n", cdpLifecycleShutdownPath)
+	mux.HandleFunc(cdpLifecycleShutdownPath, m.gracefulShutdown)
+
+	// Determine what the proxy lifecycle management server bind port is. It can be
+	// set as a flag.
+	cdpLifecycleBindPort := defaultLifecycleBindPort
+	if m.gracefulPort != 0 {
+		cdpLifecycleBindPort = strconv.Itoa(m.gracefulPort)
+	}
+	m.lifecycleServer = &http.Server{
+		Addr:    fmt.Sprintf("%s:%s", cdpLifecycleBindAddr, cdpLifecycleBindPort),
+		Handler: mux,
+	}
+
+	// Start the proxy lifecycle management server
+	go m.startLifecycleServer()
+
+	return nil
+}
+
+// startLifecycleServer starts the main proxy lifecycle management server that
+// exposes HTTP endpoints for proxy lifecycle control.
+func (m *lifecycleConfig) startLifecycleServer() {
+	m.logger.Info("starting proxy lifecycle management server", "address", m.lifecycleServer.Addr)
+	err := m.lifecycleServer.ListenAndServe()
+	if err != nil && err != http.ErrServerClosed {
+		m.logger.Error("failed to serve proxy lifecycle managerments requests", "error", err)
+		close(m.errorExitCh)
+	}
+}
+
+// stopLifecycleServer stops the consul dataplane proxy lifecycle server
+func (m *lifecycleConfig) stopLifecycleServer() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	m.running = false
+
+	if m.lifecycleServer != nil {
+		m.logger.Info("stopping the lifecycle management server")
+		err := m.lifecycleServer.Close()
+		if err != nil {
+			m.logger.Warn("error while closing lifecycle server", "error", err)
+			close(m.errorExitCh)
+		}
+	}
+}
+
+// lifecycleServerExited is used to signal that the lifecycle server
+// recieved a signal to initiate shutdown.
+func (m *lifecycleConfig) lifecycleServerExited() <-chan struct{} {
+	return m.errorExitCh
+}
+
+// gracefulShutdown blocks until shutdownGracePeriod seconds have elapsed, and, if
+// configured, will drain inbound connections to Envoy listeners during that time.
+func (m *lifecycleConfig) gracefulShutdown(rw http.ResponseWriter, _ *http.Request) {
```

Comment:
Q: Just to think through it, what if this is called multiple times concurrently?

I think from what I can tell it's okay. And we probably aren't too concerned about concurrent local requests to this endpoint anyway.

Reply:
Hooks in kubernetes can be called several times so we should definitely think about that scenario a little bit. The same will be true of the graceful startup side of things.

Reply:
This _should_ be safe as both the `/drain_listeners` and `/quitquitquit` endpoints in Envoy are idempotent:
- https://github.com/envoyproxy/envoy/blob/3b7328f253c9cdd82dc81b736a85aeaa622262dd/source/server/admin/listeners_handler.cc#L26-L27
- https://github.com/envoyproxy/envoy/blob/3b7328f253c9cdd82dc81b736a85aeaa622262dd/source/server/admin/server_cmd_handler.cc#L24
- https://github.com/envoyproxy/envoy/blob/3b7328f253c9cdd82dc81b736a85aeaa622262dd/source/server/server.cc#L1006

Double-entry should be additionally protected by the state machine transition guards in `Drain()` and `Stop()` implemented here.

Would it be alright to plan on scoping multiple call tests to future work for job handling which will add the `preStop` hook?

Reply:
Sounds good to me!